# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)(value to pass in to the lambda)"</p>

#### Syntax

In [3]:
def add_two(x):
    return x+ 2

print(add_two(4))

#lambda function syntax

a_var = 6
print((lambda x: x+2)(a_var))

6
8


#### Saving to a Variable

In [4]:
f_test = lambda x:x+2
f_test(4)

6

#### Multiple Inputs

In [6]:
#multiple inputs with no variable
print((lambda x, y, z : x * y * z)(3,5,8))

x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [7]:
#from functional programming

def multiply(f, num):
    '''
        f expects a lambda function
        num expects a number
    
    '''
    return f(num)
multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [10]:
#  normal defined function
def multiply_test(num):
    return num * 4

# function within a function
def return_func():
    test = 4
    def multiply(num):
        return num*2
    return multiply

f_return = return_func()
print(return_func())
print(f_return(4))

#lambda function return from a regular function
def return_lam(b,c):
    return lambda x,a: x + a + b + c
r_lam = return_lam(4,5)
print(r_lam(5,5))

<function return_func.<locals>.multiply at 0x000001AFB74534C0>
8
19


#### If Statements within Lambdas

In [11]:
# lambda x: true if (condition) else false
f_condition = lambda num: num * 2 if num >10 else num+2

print(f_condition(8))
print(f_condition(15))

10
30


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [1]:
f = lambda arg: arg ** 2
print(f(2))
print(f(8))

4
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [3]:
# map(function, iterable)
# iterable is something you can loop over: list dict tuple string etc
# normally the usage of map happens with a pre-defined function, but we can use lambdas as well
def squared(num, num2):
    if num < 10 and num2 <10:
        return num ** 2, num2 ** 2
    else:
        return num, num2
    
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [15]:
# map(lambda x: do something, iterable)
# uing lambda in map usually all takes place on one line
squared_nums_lam = list(map(lambda x, y: (x ** 2, y ** 2) if x< 10 and y <10 else (x,y), numbers, more_nums))
print(squared_nums_lam)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [4]:
double_nums = list(map(lambda x: x*2 - 1, numbers))
print(double_nums)
more_double_nums = list(map(lambda x: x*2 - 1, more_nums))
print(more_double_nums)

[7, 21, 39, 5, 29, 39]
[7, 19, 5, 3, 11]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [16]:
names = ['Bob', 'Andy', 'Mac', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [17]:
new_names_lam = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lam)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [6]:
from statistics import mean
# mean is sum of all values in an entity divided by the number of values in the entity

nums = [2,7,4.2,1.6,9,4.4,4.9]
nums_mean = mean(nums)
print(nums_mean)
below_mean = list(filter(lambda n: True if n > nums_mean else False, nums))
print(below_mean)

4.728571428571429
[7, 9, 4.9]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
# "just use a for loop" - Brandon

#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [8]:
# unsure if I need to do this exercise, thought I would get it a try anyway
from functools import reduce

red_num_list = [2,3,4,5,6]
correct = 2*3*4*5*6

multiplied_num = reduce(lambda a, b: a*b, red_num_list)

print(f"the correct answer is {correct}")
print(multiplied_num)

the correct answer is 720
720


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [21]:
def add_nums(num):
    #set a base case to stop recursion from going forever
    if num <=1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num-1})')
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [22]:
#5! = 5*4*3*2*1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)

factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [10]:
def sub_nums(num):
    if num <=1:
        print('sub_nums(1) = 1')
        return num
    else:
        print(f'sub_nums({num}) = {num} - sub_nums({num-1})')
        return num - sub_nums(num - 1)
    
sub_nums(5)

sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [23]:
def my_range(stop, start = 0, step = 2):
    while start < stop:
        yield start
        start += step
        
for i in my_range(20):
    my_generator_value = i
    print(my_generator_value)

0
2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [11]:
#The language of the prompt does not make sense I will be honest, until zero is reached? through both
# the squaring or squarerooting of a number you cant reach zero, and for each number?

# i modified the above function to square instead of just increment, hope that works

def squaring(stop, start = 0, step = 1):
    while start < stop:
        yield start ** 2
        start += step
        
for i in squaring(20):
    my_generator_value = i
    print(my_generator_value)

0
1
4
9
16
25
36
49
64
81
100
121
144
169
196
225
256
289
324
361


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [16]:
import re
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filled_places = list(filter(lambda name: re.match('\w+', name), places))

print(filled_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [29]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = lambda n: n.split()[-1].lower())

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [39]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

farh_temps = list(map(lambda c:  c[1]*1.8 +32, places))

farh_places = []
for i in range(len(places)):
    farh_places.append((places[i][0], farh_temps[i]))

print(farh_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [45]:
def fibonacci_printer(iteration):
    def fibs(n):
        if n == 0:
            return 1
        elif n == 1:
            return 1
        else:
            return fibs(n-1) + fibs(n-2)
    
    for i in range(iteration+1):
        print(f'Iteration {i}: {fibs(i)}')
    
fibonacci_printer(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
